embassy-rp/src/pio/mod.rs

```diff
@@ -740,6 +764,100 @@ impl<'d, PIO: Instance + 'd, const SM: usize> StateMachine<'d, PIO, SM> {
+                        Some(config.pins.sideset_base)
+                    } else {
+                        None
+                    },
+                    if config.pins.set_count > 0 {
+                        Some(config.pins.set_base)
+                    } else {
+                        None
+                    },
+                    if config.pins.out_count > 0 {
+                        Some(config.pins.out_base)
+                    } else {
+                        None
+                    },
+                ) {
+                    (None, None, None, None) => false,
+
+                    (Some(..32), None, None, None) => false,
+                    (None, Some(..32), None, None) => false,
+                    (None, None, Some(..32), None) => false,
+                    (None, None, None, Some(..32)) => false,
+
+                    (Some(..32), Some(..32), None, None) => false,
+                    (None, Some(..32), Some(..32), None) => false,
+                    (None, None, Some(..32), Some(..32)) => false,
+                    (Some(..32), None, None, Some(..32)) => false,
+
+                    (None, Some(..32), Some(..32), Some(..32)) => false,
+                    (Some(..32), None, Some(..32), Some(..32)) => false,
+                    (Some(..32), Some(..32), None, Some(..32)) => false,
+                    (Some(..32), Some(..32), Some(..32), None) => false,
+
+                    (Some(..32), Some(..32), Some(..32), Some(..32)) => false,
+
+                    (Some(16..), None, None, None) => true,
+                    (None, Some(16..), None, None) => true,
+                    (None, None, Some(16..), None) => true,
+                    (None, None, None, Some(16..)) => true,
+
+                    (Some(16..), Some(16..), None, None) => true,
+                    (None, Some(16..), Some(16..), None) => true,
+                    (None, None, Some(16..), Some(16..)) => true,
+                    (Some(16..), None, None, Some(16..)) => true,
+
+                    (None, Some(16..), Some(16..), Some(16..)) => true,
+                    (Some(16..), None, Some(16..), Some(16..)) => true,
+                    (Some(16..), Some(16..), None, Some(16..)) => true,
+                    (Some(16..), Some(16..), Some(16..), None) => true,
+
+                    (Some(16..), Some(16..), Some(16..), Some(16..)) => true,
+
+                    (i, side, set, out) => panic!(
+                        "All pins must either be <=31 or >=16, in:{:?}, side:{:?}, set:{:?}, out:{:?}",
+                        i, side, set, out
+                    ),
+                }
+            };
+
+            if shift_gpio_base {
+                sm.pinctrl().write(|w| {
+                    w.set_sideset_count(config.pins.sideset_count);
+                    w.set_set_count(config.pins.set_count);
+                    w.set_out_count(config.pins.out_count);
+                    w.set_in_base(config.pins.in_base - 16);
```

Comment:
this operation breaks if `in_base` is not set at all. Which might happen if one for example only needs sideset pins. It's also the same for all the other pin_bases. I guess there should be an additional check

Reply:
I don't think this is a problem. If one of the bases isn't being used, garbage will go into that field in the register, but if that base isn't being used it doesn't matter whats in that field.

Reply:
That ends up being a panic. Otherwise I would not have said something.
```0.042804 ERROR panicked at /root/.cargo/git/checkouts/embassy-ae7d843df5da8929/6dbab7b/embassy-rp/src/pio/mod.rs:841:35:
attempt to subtract with overflow```

Reply:
Ah, of course. I'm so used to building with release that I didn't think about it.